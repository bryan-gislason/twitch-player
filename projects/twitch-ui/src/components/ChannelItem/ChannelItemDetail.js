import PropTypes from "prop-types";
import React, { memo } from "react";
import { Image, StyleSheet, View } from "react-native";
import { useTheme } from "../../hooks";
import Text from "../Text";

const ChannelItemDetail = ({
  title,
  broadcasterName,
  broadcasterImageURL,
  categoryName,
  width: previewImageWidth,
  isHover,
  testID = "channel-item-detail",
}) => {
  const { colors, layout, typography } = useTheme();

  const rootStyle = {
    paddingHorizontal: layout.gapMedium,
    paddingTop: layout.gapSmall,
    paddingBottom: layout.gapMedium,
    width: previewImageWidth,
  };

  const broadcasterImageStyle = {
    height: typography.sizeSecondary * 2.375,
    width: typography.sizeSecondary * 2.375,
  };
  const primaryTextStyle = [
    styles.primaryText,
    {
      color: isHover ? colors.itemPrimaryHover : colors.itemPrimary,
      fontSize: typography.sizeSecondary,
    },
  ];
  const secondaryTextStyle = {
    color: isHover ? colors.itemSecondaryHover : colors.itemSecondary,
    fontSize: typography.sizeTertiary,
  };

  return (
    <View style={rootStyle} testID={testID}>
      <Text numberOfLines={1} style={primaryTextStyle}>
        {title}
      </Text>
      <View style={styles.secondaryBlock}>
        <Image
          source={{ uri: broadcasterImageURL }}
          accessibilityLabel=""
          style={broadcasterImageStyle}
          testID={`${testID}-broadcaster`}
        />
        <View style={{ marginStart: layout.gapSmall }}>
          <Text numberOfLines={1} style={secondaryTextStyle}>
            {broadcasterName}
          </Text>
          <Text numberOfLines={1} style={secondaryTextStyle}>
            {categoryName || " "}
          </Text>
        </View>
      </View>
    </View>
  );
};

ChannelItemDetail.propTypes = {
  title: PropTypes.string.isRequired,
  broadcasterName: PropTypes.string.isRequired,
  broadcasterImageURL: PropTypes.string.isRequired,
  categoryName: PropTypes.string,
  width: PropTypes.number.isRequired,
  isHover: PropTypes.bool,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  primaryText: {
    fontWeight: "bold",
  },
  secondaryBlock: {
    flexDirection: "row",
  },
});

export default memo(ChannelItemDetail);
