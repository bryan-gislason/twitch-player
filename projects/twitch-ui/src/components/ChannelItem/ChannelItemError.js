import { useTranslation } from "@twitch-player/i18n";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import { useTheme } from "../../hooks";
import Text from "../Text";

const ChannelItemError = ({
  width: previewImageWidth,
  height: previewImageHeight,
  testID = "channel-item-error",
}) => {
  const { t } = useTranslation();
  const { colors, layout, typography } = useTheme();

  const imageStyle = {
    backgroundColor: colors.itemBackground,
    height: previewImageHeight,
    width: previewImageWidth,
  };
  const detailStyle = {
    paddingHorizontal: layout.gapMedium,
    paddingVertical: layout.gapSmall,
    width: previewImageWidth,
  };
  const detailPrimaryStyle = [
    styles.detailPrimary,
    {
      color: colors.itemPrimary,
      fontSize: typography.sizeSecondary,
    },
  ];
  const detailSecondaryStyle = {
    color: colors.itemSecondary,
    fontSize: typography.sizeTertiary,
  };

  return (
    <View testID={testID}>
      <View style={imageStyle} testID={`${testID}-image`} />
      <View style={detailStyle} testID={`${testID}-detail`}>
        <Text numberOfLines={1} style={detailPrimaryStyle}>
          {t("errors.failedToRender")}
        </Text>
        <Text numberOfLines={1} style={detailSecondaryStyle}>
          {" "}
        </Text>
        <Text numberOfLines={1} style={detailSecondaryStyle}>
          {" "}
        </Text>
      </View>
    </View>
  );
};

ChannelItemError.propTypes = {
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  detailPrimary: {
    fontWeight: "bold",
    textTransform: "uppercase",
  },
});

export default memo(ChannelItemError);
