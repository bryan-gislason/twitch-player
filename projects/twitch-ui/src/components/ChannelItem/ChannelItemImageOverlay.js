import { useTranslation } from "@twitch-player/i18n";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import { useTheme } from "../../hooks";
import LiveIndicator from "../LiveIndicator";
import Text from "../Text";

const ChannelItemOverlay = ({
  viewersCount,
  style = [],
  testID = "channel-item-image-overlay",
}) => {
  const { t } = useTranslation();
  const { colors, layout, typography } = useTheme();

  const rootStyle = [styles.root].concat(style);
  const textStyle = [
    styles.text,
    {
      color: colors.liveIndicator,
      backgroundImage: `linear-gradient(transparent, ${
        colors.liveIndicatorBackground
      })`,
      paddingBottom: layout.gapSmall,
      paddingHorizontal: layout.gapMedium,
      paddingTop: layout.gapMedium,
      fontSize: typography.sizeTertiary,
    },
  ];

  return (
    <View style={rootStyle} testID={testID}>
      <LiveIndicator />
      <Text numberOfLines={1} style={textStyle}>
        {t("components.item.viewersCount", {
          count: viewersCount,
        })}
      </Text>
    </View>
  );
};

ChannelItemOverlay.propTypes = {
  viewersCount: PropTypes.number.isRequired,
  style: PropTypes.any,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  root: {
    flex: 1,
    justifyContent: "space-between",
  },
});

export default memo(ChannelItemOverlay);
