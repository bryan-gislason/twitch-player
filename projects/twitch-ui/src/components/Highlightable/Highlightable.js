import PropTypes from "prop-types";
import { Children, cloneElement, useCallback, useState } from "react";
import { isHoverEnabled } from "../../utils";

const Highlightable = ({ children }) => {
  const [isFocused, setFocused] = useState(false);
  const [isHovered, setHovered] = useState(false);
  const [isHoverable, setHoverable] = useState(true);

  const handleFocus = useCallback(() => {
    setFocused(true);
  });

  const handleBlur = useCallback(() => {
    setFocused(false);
  });

  const handleMouseEnter = useCallback(() => {
    if (isHoverEnabled()) {
      setHovered(true);
    }
  });

  const handleMouseLeave = useCallback(() => {
    setHovered(false);
  });

  // istanbul ignore next
  const handleGrant = useCallback(() => {
    setHoverable(false);
  });

  // istanbul ignore next
  const handleRelease = useCallback(() => {
    setHoverable(true);
  });

  const child = children(isFocused || (isHoverable && isHovered));

  return cloneElement(Children.only(child), {
    onFocus: handleFocus,
    onBlur: handleBlur,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onPressIn: handleGrant,
    onPressOut: handleRelease,
    onResponderGrant: handleGrant,
    onResponderRelease: handleRelease,
  });
};

Highlightable.propTypes = {
  children: PropTypes.func.isRequired,
};

export default Highlightable;
