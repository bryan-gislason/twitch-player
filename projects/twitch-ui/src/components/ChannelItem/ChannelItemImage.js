import PropTypes from "prop-types";
import React, { memo } from "react";
import { ImageBackground } from "react-native";
import { useTheme } from "../../hooks";
import { processImageTemplate } from "../../utils";
import ChannelItemImageOverlay from "./ChannelItemImageOverlay";

const ChannelItemImage = ({
  url,
  title,
  viewersCount,
  width: previewImageWidth,
  height: previewImageHeight,
  testID = "channel-item-image",
}) => {
  const { colors, layout } = useTheme();

  const imageURL = processImageTemplate(
    url,
    layout.maxChannelWidth,
    layout.maxChannelHeight
  );

  const imageStyle = {
    backgroundColor: colors.itemBackground,
    height: previewImageHeight,
    width: previewImageWidth,
  };

  return (
    <ImageBackground
      source={{ uri: imageURL }}
      accessibilityLabel={title}
      style={imageStyle}
      testID={testID}
    >
      <ChannelItemImageOverlay
        viewersCount={viewersCount}
        testID={`${testID}-overlay`}
      />
    </ImageBackground>
  );
};

ChannelItemImage.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  viewersCount: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  testID: PropTypes.string,
};

export default memo(ChannelItemImage);
