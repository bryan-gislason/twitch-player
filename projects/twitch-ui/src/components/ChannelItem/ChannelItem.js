import { ChannelType } from "@twitch-player/data";
import PropTypes from "prop-types";
import React, { memo } from "react";
import { StyleSheet, View } from "react-native";
import { useTheme } from "../../hooks";
import Highlightable from "../Highlightable";
import InternalTouchableLink from "../InternalTouchableLink";
import ChannelItemDetail from "./ChannelItemDetail";
import ChannelItemImage from "./ChannelItemImage";

const ChannelItem = ({
  item: {
    broadcaster: {
      login: broadcasterLogin,
      displayName: broadcasterName,
      profileImageURL: broadcasterImageURL,
      broadcastSettings: { title },
    },
    category,
    previewImageURL,
    viewersCount,
  },
  width: previewImageWidth,
  height: previewImageHeight,
  testID = "channel-item",
}) => {
  const { colors } = useTheme();

  const categoryName = category && category.name;

  return (
    <Highlightable>
      {isHighlighted => {
        const wrapperStyle = {
          backgroundColor: isHighlighted
            ? colors.itemBackgroundHover
            : "transparent",
        };

        return (
          <InternalTouchableLink
            href={`/channels/${broadcasterLogin}`}
            style={styles.root}
            testID={testID}
          >
            <View style={wrapperStyle} testID={`${testID}-wrapper`}>
              <ChannelItemImage
                url={previewImageURL}
                title={title}
                viewersCount={viewersCount}
                width={previewImageWidth}
                height={previewImageHeight}
                testID={`${testID}-image`}
              />
              <ChannelItemDetail
                title={title}
                broadcasterName={broadcasterName}
                broadcasterImageURL={broadcasterImageURL}
                categoryName={categoryName}
                width={previewImageWidth}
                isHighlighted={isHighlighted}
                testID={`${testID}-detail`}
              />
            </View>
          </InternalTouchableLink>
        );
      }}
    </Highlightable>
  );
};

ChannelItem.propTypes = {
  item: ChannelType.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  testID: PropTypes.string,
};

const styles = StyleSheet.create({
  root: {
    outline: "none",
  },
});

export default memo(ChannelItem);
