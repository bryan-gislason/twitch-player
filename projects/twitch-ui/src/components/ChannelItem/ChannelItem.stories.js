import { storiesOf } from "@storybook/react";
import React from "react";
import ChannelItem from ".";

const { data } = require("@twitch-player/data/fixtures/topChannels.json");
const [{ node: channel }] = data.channels.edges;

storiesOf("Static Components|ChannelItem", module)
  .add("default", () => <ChannelItem item={channel} width={400} height={225} />)
  .add("with error", () => (
    <ChannelItem item={null} width={400} height={225} />
  ));
