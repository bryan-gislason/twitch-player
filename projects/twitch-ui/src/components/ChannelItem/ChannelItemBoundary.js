import React from "react";
import ErrorBoundary from "../ErrorBoundary";
import ChannelItem from "./ChannelItem";
import ChannelItemError from "./ChannelItemError";

const ChannelItemBoundary = props => {
  return (
    <ErrorBoundary renderError={() => <ChannelItemError {...props} />}>
      <ChannelItem {...props} />
    </ErrorBoundary>
  );
};

export default ChannelItemBoundary;
