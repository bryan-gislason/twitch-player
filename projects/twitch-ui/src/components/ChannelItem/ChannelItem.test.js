import React from "react";
import { render } from "../../testing";
import Highlightable from "../Highlightable";
import ChannelItem from "./ChannelItem";

jest.mock("../Highlightable", () => jest.fn());
jest.mock("./ChannelItemDetail", () => "mock-channel-item-detail");
jest.mock("./ChannelItemImage", () => "mock-channel-item-image");

const { data } = require("@twitch-player/data/fixtures/topChannels.json");
const [{ node: channel }] = data.channels.edges;

describe("ChannelItem", () => {
  let instance;

  describe("when not hovering", () => {
    beforeEach(() => {
      Highlightable.mockImplementation(({ children }) => children(false));

      instance = render(
        <ChannelItem item={channel} width={400} height={225} />
      );
    });

    it("renders a channel", () => {
      expect(instance.getByTestId("channel-item")).toMatchSnapshot();
    });
  });

  describe("when hovering", () => {
    beforeEach(() => {
      Highlightable.mockImplementation(({ children }) => children(true));

      instance = render(
        <ChannelItem item={channel} width={400} height={225} />
      );
    });

    it("renders a channel", () => {
      expect(instance.getByTestId("channel-item")).toMatchSnapshot();
    });
  });

  describe("when the category is null", () => {
    beforeEach(() => {
      Highlightable.mockImplementation(({ children }) => children(true));

      instance = render(
        <ChannelItem
          item={{ ...channel, category: null }}
          width={400}
          height={225}
        />
      );
    });

    it("renders a channel without category name", () => {
      expect(instance.getByTestId("channel-item")).toMatchSnapshot();
    });
  });
});
