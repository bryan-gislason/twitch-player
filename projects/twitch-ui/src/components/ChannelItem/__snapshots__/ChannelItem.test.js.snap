// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelItem when hovering renders a channel 1`] = `
<div
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-1adg3ll rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-outline-1exrs8c rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
  data-focusable="true"
  data-testid="channel-item"
  style="background-color: rgb(50, 47, 55);"
  tabindex="0"
>
  <mock-channel-item-image
    height="225"
    testid="channel-item-image"
    title="Aimbot-Up Cup Duos w/ Chokzy"
    url="channels/live_user_tfue-800x450.jpg"
    viewerscount="56834"
    width="400"
  />
  <mock-channel-item-detail
    broadcasterimageurl="profiles/2470b5c6-a737-4ba6-8987-c28e0ca839e1-profile_image-150x150.jpg"
    broadcastername="Tfue"
    categoryname="Fortnite"
    ishighlighted="true"
    testid="channel-item-detail"
    title="Aimbot-Up Cup Duos w/ Chokzy"
    width="400"
  />
</div>
`;

exports[`ChannelItem when not hovering renders a channel 1`] = `
<div
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-1adg3ll rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-outline-1exrs8c rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
  data-focusable="true"
  data-testid="channel-item"
  style="background-color: rgba(0, 0, 0, 0);"
  tabindex="0"
>
  <mock-channel-item-image
    height="225"
    testid="channel-item-image"
    title="Aimbot-Up Cup Duos w/ Chokzy"
    url="channels/live_user_tfue-800x450.jpg"
    viewerscount="56834"
    width="400"
  />
  <mock-channel-item-detail
    broadcasterimageurl="profiles/2470b5c6-a737-4ba6-8987-c28e0ca839e1-profile_image-150x150.jpg"
    broadcastername="Tfue"
    categoryname="Fortnite"
    ishighlighted="false"
    testid="channel-item-detail"
    title="Aimbot-Up Cup Duos w/ Chokzy"
    width="400"
  />
</div>
`;

exports[`ChannelItem when the category is null renders a channel without category name 1`] = `
<div
  class="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-1adg3ll rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-outline-1exrs8c rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
  data-focusable="true"
  data-testid="channel-item"
  style="background-color: rgb(50, 47, 55);"
  tabindex="0"
>
  <mock-channel-item-image
    height="225"
    testid="channel-item-image"
    title="Aimbot-Up Cup Duos w/ Chokzy"
    url="channels/live_user_tfue-800x450.jpg"
    viewerscount="56834"
    width="400"
  />
  <mock-channel-item-detail
    broadcasterimageurl="profiles/2470b5c6-a737-4ba6-8987-c28e0ca839e1-profile_image-150x150.jpg"
    broadcastername="Tfue"
    ishighlighted="true"
    testid="channel-item-detail"
    title="Aimbot-Up Cup Duos w/ Chokzy"
    width="400"
  />
</div>
`;
